internal/cmd/cli/cleanup/cleanup.go

```diff
@@ -0,0 +1,113 @@
+	roleBinding := c.RBAC.RoleBinding()
+	clusterRoleBinding := c.RBAC.ClusterRoleBinding()
+
+	// lookup for all existing clusters
+	seen := map[types.NamespacedName]struct{}{}
+	clusterList, _ := cluster.List("", metav1.ListOptions{})
+	for _, c := range clusterList.Items {
+		clusterKey := types.NamespacedName{Namespace: c.Namespace, Name: c.Name}
+		seen[clusterKey] = struct{}{}
+	}
+
+	crList, _ := clusterRegistration.List("", metav1.ListOptions{})
+
+	logrus.Debugf("Found %d clusters and %d cluster registrations", len(clusterList.Items), len(crList.Items))
+
+	latestGranted := map[types.NamespacedName]metav1.Time{}
+	for _, cr := range crList.Items {
+		// delete cluster registrations that have no cluster
+		clusterKey := types.NamespacedName{Namespace: cr.Namespace, Name: cr.Status.ClusterName}
+		if _, found := seen[clusterKey]; !found {
+			if err := clusterRegistration.Delete(cr.Namespace, cr.Name, nil); err != nil && !apierrors.IsNotFound(err) {
+				logrus.Errorf("Failed to delete cluster registration %s/%s: %v", cr.Namespace, cr.Name, err)
+			}
+			continue
+		}
+
+		// figure out the latest granted registration request per cluster
+		ts := latestGranted[clusterKey]
+		if cr.Status.Granted && ts.Before(&cr.CreationTimestamp) {
+			latestGranted[clusterKey] = cr.CreationTimestamp
+		}
+	}
+
+	// sleep at max 10 seconds to not overload the API server
+	b := &backoff.Backoff{
+		Min:    time.Duration(opts.Min) * time.Millisecond,
+		Max:    time.Duration(opts.Max) * time.Second,
+		Factor: opts.Factor,
+		Jitter: true,
+	}
+
+	go func() {
+		// it should be safe to delete up to the latest granted registration
+		// requests after that might be in flight
+		// requests before that are outdated
+		for _, cr := range crList.Items {
+			time.Sleep(b.Duration())
+
+			clusterKey := types.NamespacedName{Namespace: cr.Namespace, Name: cr.Status.ClusterName}
+			latest := latestGranted[clusterKey]
+			if cr.Status.Granted && cr.CreationTimestamp.Before(&latest) {
+				if err := clusterRegistration.Delete(cr.Namespace, cr.Name, nil); err != nil && !apierrors.IsNotFound(err) {
+					logrus.Errorf("Failed to delete clusterregistration %s/%s: %v", cr.Namespace, cr.Name, err)
+				}
+				// also try to delete orphans
+				_ = role.Delete(cr.Namespace, cr.Name, nil)
+				_ = roleBinding.Delete(cr.Namespace, cr.Name, nil)
+
+				filter := labels.Set(map[string]string{"fleet.cattle.io/cluster-registration": cr.Name, "fleet.cattle.io/cluster-registration-namespace": cr.Namespace})
+				saList, _ := serviceAccount.List("", metav1.ListOptions{LabelSelector: filter.AsSelector().String()})
+				for _, sa := range saList.Items {
+					_ = serviceAccount.Delete(sa.Namespace, sa.Name, nil)
+				}
+
+				owner := labels.Set(map[string]string{"objectset.rio.cattle.io/owner-name": cr.Name, "objectset.rio.cattle.io/owner-namespace": cr.Namespace}).AsSelector().String()
+				rbList, _ := roleBinding.List("", metav1.ListOptions{LabelSelector: owner})
```

Comment:
nit, for style consistency with the above block:
```suggestion
				owner := labels.Set(map[string]string{"objectset.rio.cattle.io/owner-name": cr.Name, "objectset.rio.cattle.io/owner-namespace": cr.Namespace})
				rbList, _ := roleBinding.List("", metav1.ListOptions{LabelSelector: owner.AsSelector().String()})
```